Body Library Refactoring

akeeste opened the pull request:
Addresses part of the project card here: https://github.com/orgs/WEC-Sim/projects/12/views/1?pane=issue&itemId=91263089

This PR refactors the body library:
- breaks out the different body types into multiple .slx files
- removes the nonhydro body. Use the drag body instead
- more closely links the flex and rigid body where possible
- removes the WEC-Sim GUI, while retaining the ability to run from Simulink

To update applications to work with this PR:
- nonhydro and drag bodies can maintain their input file set-up but need to be replaced with a Drag Body block in the model file.
- broken library links updated for the OWC orifice model case

TODO
- [x] update documentation
- [x] remove flex body GUI
- [x] get the mask preview ("body(*)") working for drag and flex bodies
- [x] replace the previous body library with a forwarding table that points to the new library files
- [x] corresponding applications repo PR https://github.com/WEC-Sim/WEC-Sim_Applications/pull/88
- [ ] mask initialization functions to separate files --> save this for another PR

akeeste commented:
@kmruehl this PR is ready for review. I was able to get the few top-level mask initialization functions moved to .m files, but not the rest. I believe all of the applications cases run. The regression tests fail but I don't know their most recent status.

I was not able to update documentation yet to reflect the removal of the nonhydro body

kmruehl commented:
Thanks @akeeste I'm making a few revisions to the library:
- updating all library files to 2020b
- naming all libraries so that they are alphabetically grouped, e.g., `lib_body_drag` and `lib_body_hydro`, instead of `lib_drag_body` and `lib_hydro_body`
- resolving the linking between the main library and the bodies library

kmruehl commented on tests/runFromSimTest.m:
I understand that we're removing the WEC-Sim GUI, but I think that we should still test the run from Simulink functionality. Why was this test removed?

akeeste replied:
The test was called "run from Simulink" but was primarily a test of the WEC-Sim GUI. We could add this test back and heavily simplify it to only test running WEC-Sim from Simulink

kmruehl commented on wecSimTest.m:
```diff
@@ -34,9 +34,6 @@
-%   'runFromSimTest'   Run tests for execution of WECSim from Simulink.
-%                      Default is true.
-%
 %   'rotationTest'     Run rotation tests. Default is true.
```
Same comment as above regarding testing run from Simulink

kmruehl reviewed (commented):
@akeeste thanks for working on this. I made some minor updates that broke then library. Let's debug when you're back in the office and discuss next steps

akeeste commented:
> Thanks @akeeste I'm making a few revisions to the library:
> 
> * updating all library files to 2020b
> * naming all libraries so that they are alphabetically grouped, e.g., `lib_body_drag` and `lib_body_hydro`, instead of `lib_drag_body` and `lib_hydro_body`
> * resolving the linking between the main library and the bodies library

also for some reason saving the library files back to 2020b seems to remove the mask initialization that has been moved to `.m` files, so blocks are not initializing (hence the duplicate goto tags)

akeeste commented:
TODO after @akeeste and @kmruehl discussion:
- [ ] combine renaming and restructuring from @kmruehl's last commit with the inclusion of .m mask initializations 
- [ ] evaluate for what version saving backwards gets rid of .m file mask initialization

akeeste commented:
@kmruehl I updated the library structure, fixed the links, and appear to have things working again. Next time I can work on WEC-Sim I'll document the library changes in this PR. Likewise, can you comment on what changed in the non-body library files in commit 2f7f460?

kmruehl commented:
@akeeste is this ready for review?

kmruehl commented:
I'm still having issues with the Simulink Library rendering. I'm running MATLAB 2024a. Maybe someone else on the team can check to see if they're having the same issue?

<img width="606" alt="image" src="https://github.com/user-attachments/assets/42729042-0ef0-4c01-81de-59b662f00432" />

akeeste commented:
@kmruehl Yes this is ready for review.

I'm not sure what's causing the library appearance issue. On R2024b PTO-Sim renders fine but not the rest. I agree that maybe it's a MATLAB version problem

kmruehl commented:
@akeeste that's the same for me with 2024a. PTO-Sim renders properly, but nothing else does. 

Everything renders properly for me on 2025a and @jtgrasb confirmed that 2023b renders properly, so it must be an issue with the 2024a and 2024b distributions. I don't think it's worth debugging because the library blocks still work, they just don't render in the library. I'm just mentioning it in this thread for future awareness.

kmruehl commented on docs/user/code_structure.rst:
```diff
@@ -409,8 +409,8 @@ following lines in the WEC-Sim input file, where ``i`` is the body number and
-drag, or nonhydrodynamic. These types represent varying degrees of complexity
+WEC-Sim bodies may be one of three types\: hydrodynamic and rigid, hydrodynamic and flexible, 
+or nonhydrodynamic and rigid (drag body). These types represent varying degrees of complexity
 and require various input parameters and BEM data, detailed in the table below.
```
There are some typos here that I will made a commit to resolve

akeeste replied:
Feel free to clarify. There's a few overlapping features when we talk about the types of bodies (rigid vs flex, hydro vs nonhydro)

akeeste replied:
Note the code change showing above is outdated and for some reason GitHub is not labeling it as such

akeeste replied:
That's detailed in this section: https://wec-sim.github.io/WEC-Sim/main/user/code_structure.html#body-blocks. 

The rest of the docs Code Structure section have an identical format, first breaking out each class and then the respective objects. Though any differences here go away once the rigid and flex are fully linked

kmruehl commented on docs/user/advanced_features.rst:
```diff
@@ -65,48 +65,6 @@ Running from Simulink
-	* Type ``initializeWecSim`` in the Command Window
-	* Run the model from Simulink
-	* Wait for the simulation to complete, then type ``stopWecSim`` in the Command Window
-* Run from Simulink with custom parameters
```
It seems like this as good information that has now been removed. Shouldn't we still explain to users how to run the Run from Simulink feature?

kmruehl replied:
Thanks @akeeste, I assume you're going to make this change?

akeeste replied:
done

kmruehl commented on source/functions/initializeWecSim.m:
Shouldn't lines 77-83 be removed?

```matlab 

elseif strcmp(values{j},'Custom Parameters')
        % wecSim input from custom parameters in Simulink block
        fprintf('\nWEC-Sim Input From Custom Parameters In Simulink... \n');
        inputFile = 'wecSimInputFile_simulinkCustomParameters';
        writeInputFromBlocks(inputFile);
        run(inputFile);
    end

```

kmruehl commented on source/functions/initializeWecSim.m:
```diff
@@ -532,20 +503,12 @@
-end; clear i
+fprintf('\nNumber of Hydro Bodies = %u \n',simu.numHydroBodies)
+fprintf('\nNumber of Drag Bodies = %u \n',numDragBodies)
+for i = 1:length(body)
```
Number of drag bodies should also be removed

akeeste replied:
The number of drag bodies is still relevant. We previously enumerated hydro, nonhydro, and drag bodies. Now we just enumerate hydro and drag

kmruehl commented on docs/user/code_structure.rst:
```diff
@@ -434,15 +434,6 @@ important distinctions.
 |                         |``body(i).volume``                           |
 |                         |``body(i).nonHydro=1``                       |
 +-------------------------+---------------------------------------------+
```
I know this isn't rendering, but we should update it and resolve the rendering issue

akeeste replied:
What is the current rendering issue?

kmruehl replied:
I was referring to the note on line 418 `.. TO DO: This table is not rendering properly`. It looks like the table is rendering properly, so that note should be removed.

akeeste replied:
I'm guessing the table was intended render each code line as a separate line in the table but it does not:  
<img width="551" height="172" alt="image" src="https://github.com/user-attachments/assets/c0fcb5a7-295a-4d2e-987d-a5bc9a7f8962" />

kmruehl commented on source/objects/bodyClass.m:
```diff
@@ -59,7 +59,7 @@
         name (1,:) {mustBeText}                     = ''                    % (`string`) Specifies the body name. For hydrodynamic bodies this is defined in h5 file. For nonhydrodynamic bodies this is defined by the user, Default = ``[]``.
-        nonHydro (1,1) {mustBeInteger}              = 0                     % (`integer`) Flag for non-hydro body, Options: 0 (hydro body), 1 (non-hydro body), 2 (drag body). Default = ``0``.
+        nonHydro (1,1) {mustBeInteger}              = 0                     % (`integer`) Flag for non-hydro body, Options: 0 (hydro body), 1 or 2 (drag body). Default = ``0``.
         nonlinearHydro (1,1) {mustBeInteger}        = 0                     % (`integer`) Flag for nonlinear hydrohanamics calculation, Options: 0 (linear), 1 (nonlinear), 2 (nonlinear). Default = ``0``
```
What is the difference between options 1 and 2? Isn't one a drag body and the other a non-hydro body (aka a drag body without drag)?

Also, the naming convention no longer makes sense. It's called `body.nonHydro`, but the default is for it to be a hydro body, and the options are two drag bodies?

kmruehl replied:
I understand that it doesn't affect rigid/flex now, I was referring to the future plan to have one body library block , with multiple variants for: rigid, flex, and drag. In that case, the major overhaul will happen in a future PR, so it may not make sense to remove nonHydro options 1 and 2 now and cause backward compatibility issues. 

This is also related to my above comment. Long-term, it does not make sense to have two `body.nonHydro` flags that do the same thing, i.e. `body.nonHydro = 1` and `body.nonHydro = 2`. Similarly, it does not make sense to have `body.nonHydro = 1` activate a drag body. It would make more sense to have `body.drag = 1` activate a drag body. However, long-term we will have multiple variants of this block, so we should think about what that will look like, and whether it makes sense to begin making those changes in this PR, or a future PR. 

For example, here are some options:

` body.hydro= 1` activates a hydro body (default = 1)
` body.drag = 1` activates a drag body (default = 0)
` body.flex= 1` activates a drag body (default = 0)
In this case we would need to check that multiple flags are not activated at once

`body.type` or `body.variant` = 1 for hydro body , 2 for drag body, 3 for hydro body (default = 1)
This would likely be the easiest to implement, but may not be as intuitive for users

akeeste replied:
That makes sense. Since we plan for another significant body library PR, I'm inclined to postpone the change until the flex body is addressed and it can be done all at once. How does that sound? It would be nice to get the current changes in and fine tune things in the next PR

kmruehl replied:
I agree, let's postpone the refactoring until we merge flex-rigid into body. I just wanted to start the discussion about how we plan to implement it, and decide whether it makes sense to do it now or later. Consensus on waiting until a future PR

kmruehl commented on tests/runFromSimTest.m:
```diff
@@ -11,113 +11,22 @@
-                                         'runFromSimulinkTests');
+            obj.runFromSimDir = fullfile(obj.testDir, '..', ...
+                'examples', 'RM3');
         end        
```
Are you removing the `example\RM3FromSimulink` directory, and instead running from Simulink with the `example/RM3`? That makes sense to me, but the `example/RM3` case isn't running for me directly from Simulink.

akeeste replied:
This should be fixed, there was a reference to the non-hydro body in the Rigid Body mask that should have been removed

kmruehl replied:
Thanks, confirmed that this issue is resolved!

kmruehl reviewed (changes requested):
@akeeste thanks for drafting this PR, but I think it needs some work prior to merging. Several of the docs need to be updated, and so does `initializeWecSim` and `bodyClass`. Right now I am not able to run the RM3 from Simulink, and the docs for the Body Library updates are confusing.

kmruehl commented:
@akeeste I made some very minor updates towards the requested changes. Let's chat about how to get these revisions done and merging the PR. Thanks!

akeeste commented:
Thanks @kmruehl I'll make the requested changes. I don't know why the tests are failing. They have not run consistently for a while and the regression were only recently fixed

akeeste commented:
It looks like the tests are not running to completion, I'll look into that too

akeeste commented:
Tests finally passing. Merged

kmruehl commented:
Thanks @akeeste
